Amazon.QLDB.Driver.Tests/AsyncQldbDriverTests.cs

```diff
@@ -240,93 +264,92 @@ public async Task TestAsyncExecuteWithFuncLambdaAndRetryPolicyReturnsFuncOutput(
-                {
-                    RequestId = "testId"
-                }
-            };
-        
-            var sendCommandResponseCommit = new SendCommandResponse
+            h1 = Transaction.Dot(h1, statement, new List<IIonValue> { });
+
+            mockClient.QueueResponse(startSessionResponse);
+            mockClient.QueueResponse(startTransactionResponse);
+            foreach (var ex in exceptions)
             {
-                CommitTransaction = new CommitTransactionResult
-                {
-                    TransactionId = TestTransactionId,
-                    CommitDigest = new MemoryStream(h1.Hash),
-                },
-                ResponseMetadata = new ResponseMetadata
+                mockClient.QueueResponse(ex);
+
+                // OccConflictException reuses the session so no need for another start session.
+                if (ex is not OccConflictException)
                 {
-                    RequestId = "testId"
+                    mockClient.QueueResponse(startSessionResponse);
                 }
-            };
-            var mockCreator = new Mock<Func<CancellationToken, Task<Session>>>();
-            var mockSession = new Mock<Session>(null, null, null, null, null);
-        
-            mockSession.Setup(x => x.StartTransactionAsync(default)).ReturnsAsync(sendCommandResponseWithStartSession.StartTransaction);
-            mockSession.SetupSequence(x => x.ExecuteStatementAsync(It.IsAny<String>(), It.IsAny<String>(), It.IsAny<List<IIonValue>>(), default))
-                .Throws(exception)
-                .Throws(exception)
-                .ReturnsAsync(sendCommandResponseExecute.ExecuteStatement);
-            mockSession.Setup(x => x.CommitTransactionAsync(It.IsAny<String>(), It.IsAny<MemoryStream>(), default))
-                .ReturnsAsync(sendCommandResponseCommit.CommitTransaction);
-        
-            mockCreator.Setup(x => x(default)).ReturnsAsync(mockSession.Object);
-        
-            var driver = new AsyncQldbDriver(
-                new AsyncSessionPool(mockCreator.Object, AsyncQldbDriverBuilder.CreateDefaultRetryHandler(NullLogger.Instance), 4, NullLogger.Instance));
-        
+
+                mockClient.QueueResponse(startTransactionResponse);
+            }
+            mockClient.QueueResponse(executeResponse(null));
+            mockClient.QueueResponse(commitResponse(h1.Hash));
+
+            var driver = new AsyncQldbDriver("ledgerName", mockClient, 4, NullLogger.Instance);
+
             try
             {
-                await driver.Execute(txn => txn.Execute(statement));
+                await driver.Execute(txn => txn.Execute(statement), policy);
+
+                Assert.IsNull(expectedExceptionType);
             }
             catch (Exception e)
             {
-                Assert.IsTrue(expectThrow);
-                Assert.AreEqual(exception, e);
-                return;
+                Assert.IsNotNull(expectedExceptionType);
```

Comment:
The next line is testing e with `Assert.IsInstanceOfType(e, expectedExceptionType);`

The point of `Assert.IsNotNull(expectedExceptionType);` is to establish that we are supposed to be in the catch block as `expectedExceptionType` is null for test cases that should not throw exceptions (and also probably to prevent `Assert.IsInstanceOfType` from evaluating against a null).

Reply:
Oh wow, I totally missed that as the intent, can you add a comment to that effect then?

Reply:
I just reworked this so the test cases pass in an `expectThrow` flag instead of `expectedExceptionType` and then we infer the expected type by checking the type of the last exception in the list.

I think this makes the intent quite a bit more clear.